### ISSUE TITLE
MPI errors in SVN 7701 if just one fire mesh

### ISSUE THREAD

**Issue by gforney** [firemodels/fds#1301]
```
 Compilation Date : Thu, 10 Feb 2011
 Version: 6.0.0; MPI Enabled; OpenMP Disabled
 SVN Revision No. : 7701

I noticed that if there is just one fire mesh and
you are doing a MPI calculation then there are some
errors. Below are the changes that I made to my own
version so that the program did run. I did not check
if I got correct results. The only thing I checked 
was that there was no error messages for fds5_mpi_debug.
How I found that? Well, I have quite often just one fire
mesh and then some evacuation meshes, which means for
the MPI that there are two processes one for fire mesh
and one for all evacuation meshes.

Without the lines 302 and 325:

Fatal error in MPI_Send: Other MPI error, error stack:
MPI_Send(173): MPI_Send(buf=0AF771B4, count=1, MPI_INTEGER, dest=0, tag=1, MPI_C
OMM_WORLD) failed
MPID_Send(53): DEADLOCK: attempting to send a message to the local process witho
ut a prior matching receive
job aborted:
rank: node: exit code[: error message]
0: ESPKT4K163: 1: Fatal error in MPI_Send: Other MPI error, error stack:
MPI_Send(173): MPI_Send(buf=0AF771B4, count=1, MPI_INTEGER, dest=0, tag=1, MPI_C
OMM_WORLD) failed
MPID_Send(53): DEADLOCK: attempting to send a message to the local process witho
ut a prior matching receive

And without 1937 (lines 302 and 325 added):

CygWin (/cygdrive/e/SVN_Verification_Evacuation): mpiexec -file config.txt
Process   0 of   0 is running on ESPKT4K163.ad.vtt.fi
Mesh   1 is assigned to Process   0
forrtl: severe (408): fort: (2): Subscript #1 of the array REQ has value 1 which
 is greater than the upper bound of 0

Image              PC        Routine            Line        Source
fds5_mpi_debug.ex  018D764A  Unknown               Unknown  Unknown
fds5_mpi_debug.ex  018D4838  Unknown               Unknown  Unknown
fds5_mpi_debug.ex  01866EB8  Unknown               Unknown  Unknown
fds5_mpi_debug.ex  018674E1  Unknown               Unknown  Unknown
fds5_mpi_debug.ex  0181400F  _FDS_ip_MESH_EXCH        1939  main.f90
fds5_mpi_debug.ex  01792687  _MAIN__                   353  main.f90
fds5_mpi_debug.ex  0189E5F3  Unknown               Unknown  Unknown
fds5_mpi_debug.ex  0185A272  Unknown               Unknown  Unknown
kernel32.dll       7C817077  Unknown               Unknown  Unknown

job aborted:
rank: node: exit code[: error message]
0: ESPKT4K163: 408: process 0 exited without calling finalize

Below is the "diff" of my changes:

========================================
 Now comparing  main.f90 (< Own, > SVN)
========================================
303d302
<       IF (NOM==NM) CYCLE
327d325
<       IF (NOM==NM) CYCLE
1939c1937
< IF (USE_MPI .AND. CODE/=2 .AND. CODE/=5 .AND. N_REQ>0) CALL MPI_WAITALL(N_REQ,REQ(1:N_REQ),MPI_STATUSES_IGNORE,IERR)

---
> IF (USE_MPI .AND. CODE/=2 .AND. CODE/=5) CALL MPI_WAITALL(N_REQ,REQ(1:N_REQ),MPI_STATUSES_IGNORE,IERR)

TimoK
```

Original issue reported on code.google.com by `tkorhon1` on 2011-02-11 08:55:39

**Comment by gforney** [comment]
```
Thanks, Timo. I recently made changes to the MPI calls to reduce the amount of memory
that we allocate for these arrays like NIC, I_MIN, etc. Previously, these arrays were
dimensioned NMESHES by NMESHES, and if we starting using hundreds of meshes, this will
be a big waste.
```

Original issue reported on code.google.com by `mcgratta` on 2011-02-11 13:17:30

**Comment by gforney** [comment]
```
Timo -- post the input file that did not work. Thanks.
```

Original issue reported on code.google.com by `mcgratta` on 2011-02-11 13:19:45

**Comment by gforney** [comment]
```
Kevin, as just discussed per mail, here is the input file which also caused problems
in parallel mode. In serial mode, it seems to run
```

Original issue reported on code.google.com by `kiliansusan` on 2011-02-11 13:30:23

<hr>
- _Attachment: [scarc2d_fft_2mesh.fds](https://storage.googleapis.com/google-code-attachments/fds-smv/issue-1331/comment-3/scarc2d_fft_2mesh.fds)_

**Comment by gforney** [comment]
```
I made the changes Timo suggested. Two problems -- the MPI_SEND calls were sending messages
from one process to itself, which the linux executable allowed but the Windows did
not. Then in Timo's case, there was an MPI_WAITALL call when there were no MPI calls
to wait for.

Timo, Susan -- make sure your cases now work properly. Check not only that the cases
run, but also that the results are the same as before. Thanks.
```

Original issue reported on code.google.com by `mcgratta` on 2011-02-11 14:31:59

**Comment by gforney** [comment]
```
Fine, this seems to work now. On the first sight, the results seem to be the same as
before. 
```

Original issue reported on code.google.com by `kiliansusan` on 2011-02-11 14:54:00

**Comment by gforney** [comment]
```
Great, thanks for catching this. If a new error appears, just start a new issue. I will
close this one to avoid confusion.
```

Original issue reported on code.google.com by `mcgratta` on 2011-02-11 15:19:19
